dali/dfu/dfuutil.cpp

```diff
@@ -868,10 +897,169 @@ class CDFUhelper: public CInterface, implements IDFUhelper
+                        StringBuffer subquery;
+                        sublname.makeFullnameQuery(subquery, DXB_File, true);
+                        Owned<IRemoteConnection> subconn = querySDS().connect(subquery.str(),myProcessSession(),RTM_LOCK_WRITE, FOREIGN_DALI_TIMEOUT);
+                        if (!subconn)
+                        {
+                            sublname.makeFullnameQuery(subquery.clear(), DXB_SuperFile, true);
+                            subconn.setown(querySDS().connect(subquery.str(),myProcessSession(),0, FOREIGN_DALI_TIMEOUT));
+                        }
+                        if (!subconn)
+                        {
+                            ERRLOG("SuperFile %s is missing sub-file file %s",superName.str(),subfilename.str());
+                            continue;
+                        }
+                        subnum++;
+
+                        Owned<IPropertyTree> subroot = subconn->getRoot();
+                        Owned<IPropertyTreeIterator> iter = subroot->getElements("SuperOwner");
+                        StringBuffer pname;
+                        bool parentok=false;
+                        ForEach(*iter)
+                        {
+                            iter->query().getProp("@name",pname.clear());
+                            if (strcmp(pname.str(),superName.str())==0)
+                                parentok = true;
+                            else
+                            {
+                                CDfsLogicalFileName sdlname;
+                                sdlname.set(pname.str());
+                                StringBuffer sdquery;
+                                sdlname.makeFullnameQuery(sdquery, DXB_SuperFile, true);
+                                Owned<IRemoteConnection> sdconn = querySDS().connect(sdquery.str(),myProcessSession(),0, FOREIGN_DALI_TIMEOUT);
+                                if (!conn)
+                                    WARNLOG("SubFile %s has missing owner superfile %s",sublname.get(),sdlname.get());
+                            }
+                        }
+                        if (!parentok)
+                        {
+                            WARNLOG("SubFile %s is missing link to Superfile %s",sublname.get(),superName.str());
+                            Owned<IPropertyTree> t = createPTree("SuperOwner");
+                            t->setProp("@name",superName.str());
+                            subroot->addPropTree("SuperOwner",t.getClear());
+
+                        }
+                        subconn->commit();
+                        subconn->close();
+                    }
+                }
+            }
+            PROGLOG("%d of %d sub file(s) reference to superfile %s are fixed.",subnum, numSubfiles, superName.str());
+        }
+        else
+        {
+            // Logical file map
+            Owned<IRemoteConnection> conn = querySDS().connect(lfn, myProcessSession(), 0, FOREIGN_DALI_TIMEOUT);
```

Comment:
various reasons, access to Dali (or e.g. Cassandra) needs to be tightly encapsulated, all access to it should go via well defined interfaces.
dfuplus is a tool that talks via those interfaces